File: Homework-01/main.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

long double c_conv(int in_channel, long int o_channel, int kernel_size, int stride, float ***input_image);

int main(int argc, char** argv)
{
    int in_channel, kernel_size, stride;
    int i,j,k;
    long int o_channel;
    int rows, columns;
    float ***img_array;
    double computation_time[11];
    long double num_of_ops;
    clock_t start, end;

    //Image 0 size
    rows = 720;
    columns = 1280;
    
    //Image 1 size
    //rows = 1080;
    //columns = 1920;
    in_channel = 3;
    kernel_size = 3;
    stride = 1;
    
    //Allocate space for 3D image array
    img_array = (float***)malloc(in_channel*sizeof(float**));
    for (i = 0; i < in_channel; i++)
        img_array[i] = (float**)malloc(rows*sizeof(float*));
    
    for (i = 0; i < in_channel; i++) 
        for (j = 0; j < rows; j++) 
            img_array[i][j] = (float*)malloc(columns*sizeof(float));
       
    /*Create input test image*/
    for(i = 0; i < in_channel; i++)
        for(j = 0; j < rows; j++)
            for(k = 0; k < columns; k++){
                img_array[i][j][k] = rand()%255;
                //printf("\n%f",img_array[i][j][k]);
            }
                
    for(i = 0; i < 11; i++)
        computation_time[i] = 0; //Variable initialization for time taken
   
    //Initialize Conv2D
    printf("\n------------------Part D--------------------\n");
    for(i = 0; i <11; i++)      
    {
        o_channel = pow(2,i);
        start = clock();
        num_of_ops = c_conv(in_channel, o_channel, kernel_size, stride, img_array);   //Call c_conv function
        end = clock();
        computation_time[i] = (double)(end - start) / CLOCKS_PER_SEC;
        printf("i = %d, o_channel = %ld, number_of_operations = %Lf, computation_time = %lf \n", i, o_channel, num_of_ops, computation_time[i]);
    }
    return 0;

}


long double c_conv(int in_channel, long int o_channel, int kernel_size, int stride, float ***input_image)
{
    float ***kernel, ***out_img_array; //kernel and output array declaration 
    int i,j,k,a,b,c,x,y,z,roi_offset;
    int num_of_ops = 0;
    int count_mul = 0, count_add = 0;
    float mul_acc = 0;
    long double count = 0;
    
    //Image 0 size
    int rows = 720;
    int columns = 1280;
    
    //Image 1 size
    //int rows = 1080;
    //int columns = 1920;
    
    //Allocate space for 3D kernel
    kernel = (float***)malloc(kernel_size*sizeof(float**));
    for (i = 0; i < kernel_size; i++)
        kernel[i] = (float**)malloc(kernel_size*sizeof(float*));
    
    for (i = 0; i < kernel_size; i++) 
        for (j = 0; j < kernel_size; j++) 
            kernel[i][j] = (float*)malloc(kernel_size*sizeof(float));
    
    //Create 3D kernel
    for(i = 0; i < kernel_size; i++)
        for(j = 0; j < kernel_size; j++)
            for(k = 0; k < kernel_size; k++){
                kernel[i][j][k] = (float)(rand()%20-10.0)/10.0;
                //printf("\n%f",kernel[i][j][k]);
            }
          
                        
    //Allocate space for 3D output array
    out_img_array = (float***)malloc(o_channel*sizeof(float**));
    for (i = 0; i < o_channel; i++)
        out_img_array[i] = (float**)malloc(rows*sizeof(float*));
    
    for (i = 0; i < o_channel; i++) 
        for (j = 0; j < rows; j++) 
            out_img_array[i][j] = (float*)malloc(columns*sizeof(float));
       
    /*Initialize output image with all zeros*/
    for(i = 0; i < o_channel; i++)
        for(j = 0; j < rows; j++)
            for(k = 0; k < columns; k++)
                out_img_array[i][j][k] = 0;
                
    roi_offset = (int)(kernel_size/2);
    //printf("\nk size = %d, offset = %d",kernel_size,roi_offset);
    
    //Loop over every pixel of the image
    for(i = 0; i < o_channel; i++)
        for(j = roi_offset; j < rows-roi_offset; j++)
            for(k = roi_offset; k < columns-roi_offset; k++)
            {
                mul_acc = 0; //Initialize variable for each step as kernel is shifted through the image
                //Multiply and accumulate region of image of size 3x3x3 with kernel of 3x3x3
                for(x = 0, a = 0; x < kernel_size; x++, a++){
                    for(y = 0, b = j-roi_offset; y < kernel_size; y++, b++){
                        for(z = 0, c = k-roi_offset; z < kernel_size; z++, c++){
                            mul_acc += kernel[x][y][z] * input_image[a][b][c];
                            count += 2;
                            //printf("i%d, j%d, k%d, a%d, b%d, c%d, x%d, y%d, z%d, %f \n",i,j,k,a,b,c,x,y,z,mul_acc);
                        }//end for
                        //printf("\n");
                    }
                    //printf("\n\n");
                }//end mul and acc for
                out_img_array[i][j][k] = mul_acc;
             }

    return count; //num_of_ops;
}

File: Homework-01/main.py
#Main function for Assignment for Image Convolution
from conv import Conv2D
from PIL import Image
import numpy as np
import matplotlib.pyplot as plt
import torch
import torchvision
import torchvision.transforms as transforms
import time


imgToTensor = transforms.Compose([transforms.ToTensor()])   #Convert PIL.Image to tensor

num_images = 2
#Load Test Images and convert to 3D Float Tensor
input_image0 = Image.open("img0.jpg")
input_image1 = Image.open("img1.jpg")

#input_image2 = Image.open("img2.jpg")
#input_image3 = Image.open("img3.jpg")
input_img_tensor = (imgToTensor(input_image0),imgToTensor(input_image1))
#input_img_tensor = (imgToTensor(input_image0),imgToTensor(input_image1),imgToTensor(input_image2),imgToTensor(input_image3))

computation_time = np.zeros(10) #Variable declaration for Part B, ones less value than 11 as last value will be extrapolated
num_of_ops_partc = np.zeros(5)  #Variable declaration for Part C

#Steps for Part A to C for two input image tensors
for j in range(num_images):
    print("\n-------------------------IMAGE %r (%r)--------------------------\n\n"%(j, input_img_tensor[j].size()))
    print("---------------------PART A------------------------")    
    """PART A"""
    """TASK 1"""
    #Initialize Conv2D
    conv2d = Conv2D(in_channel=3, o_channel=1, kernel_size=3, stride=1, mode ='known')
    [num_of_ops, output_img_tensor] = conv2d.forward(input_img_tensor[j])   #Call forward function with input image tensor
    num_ops = '\nThe total number of operations (multiplications and additions) for Task1 for Image No %d is %d'%(j,num_of_ops)
    print (num_ops)
    output_img_array = output_img_tensor.numpy()    #Convert from Float Tensor to numpy.array
    conv2d.normalize(j,output_img_array)              #Normalize and Save as grayscale image
    
    """TASK 2"""
    #Initialize Conv2D
    conv2d = Conv2D(in_channel=3, o_channel=2, kernel_size=5, stride=1, mode ='known')
    [num_of_ops, output_img_tensor] = conv2d.forward(input_img_tensor[j])   #Call forward function with input image tensor
    num_ops = '\nThe total number of operations (multiplications and additions for Task2 for Image No %d is %d'%(j,num_of_ops)
    print (num_ops)
    output_img_array = output_img_tensor.numpy()    #Convert from Float Tensor to numpy.array
    conv2d.normalize(j,output_img_array)   
    
    """TASK 3"""
    #Initialize Conv2D
    conv2d = Conv2D(in_channel=3, o_channel=3, kernel_size=3, stride=2, mode ='known')
    [num_of_ops, output_img_tensor] = conv2d.forward(input_img_tensor[j])   #Call forward function with input image tensor
    num_ops = '\nThe total number of operations (multiplications and additions for Task3 for Image No %d is %d'%(j,num_of_ops)
    print (num_ops)
    output_img_array = output_img_tensor.numpy()    #Convert from Float Tensor to numpy.array
    conv2d.normalize(j,output_img_array)
    
    
    #### PART B ###
    #Initialize Conv2D
    print("\n------------------Part B--------------------\n")
    for i in range(0,10):   #Only do till o_channel = 512, as for o_channel = 1024, total time taken increases beyond 1 hour
        conv2d = Conv2D(in_channel=3, o_channel=2**i, kernel_size=3, stride=1, mode ='rand')
        start_time = time.time()
        [num_of_ops, output_img_tensor] = conv2d.forward(input_img_tensor[j])   #Call forward function with input image tensor
        computation_time[i] = time.time() - start_time
        print('i = %r, o_channel = %r, number_of_operations = %r, computation_time = %r'%(i,2**i,num_of_ops,computation_time[i]))
            
    i = np.arange(0,10,1)
    
    #Do extrapolation for o_channel = 1024, i = 10 using polynomial curve fitting
    z=np.polyfit(i,computation_time,3)
    func=np.poly1d(z)
    other_i = 10
    other_time = func(other_i)
    
    #Create new dataset using extrapolated value
    i = np.arange(0,11,1)
    final_computation_time = np.hstack((computation_time,other_time))

    #Plot time taken for each forward pass with respect to i
    ll = plt.plot(i,final_computation_time)
    xl = plt.xlabel('i')
    yl = plt.ylabel('Time taken ')
    ttl = plt.title('Time taken for forward pass of convolution vs i')
    grd = plt.grid(True)
    plt.show()
    fig_name = '/home/soumendu/Documents/Python /BME 595/Assignments/PlotB_TimeTaken_vs_i_image%s.jpg'%(j)
    plt.savefig(fig_name, bbox_inches = 'tight')
    plt.close()

    """
    ### PART C ###
    #Initialize Conv2D
    print("\n------------------Part C--------------------\n")
    k_size = [3,5,7,9,11]    #Note computation for kernel size 3 already done
    for i in range(0,5):
        conv2d = Conv2D(in_channel=3, o_channel=2, kernel_size=k_size[i], stride=1, mode ='rand')        
        [num_of_ops_partc[i], output_img_tensor] = conv2d.forward(input_img_tensor[j])   #Call forward function with input image tensor
        print('i = %r, o_channel = %r, kernel_size = %r, number_of_operations = %r'%(i, 2, k_size[i], num_of_ops_partc[i]))
            
    #Plot number of operations =with respect to kernel size
    ll = plt.plot(k_size,num_of_ops_partc)
    xl = plt.xlabel('kernel size')
    yl = plt.ylabel('Number of Operations')
    ttl = plt.title('Number of Operations vs Kernel Size')
    grd = plt.grid(True)
    plt.show()
    fig_name = '/home/soumendu/Documents/Python /BME 595/Assignments/PlotC_NumOps_vs_kernelsize_image%s.jpg'%(j)
    plt.savefig(fig_name, bbox_inches = 'tight')
    plt.close()
    #"""


